x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/score/Decay.java

```diff
@@ -0,0 +1,618 @@
+    static double process(
+        BytesRef value,
+        @Fixed BytesRef origin,
+        @Fixed BytesRef scale,
+        @Fixed BytesRef offset,
+        @Fixed double decay,
+        @Fixed BytesRef functionType
+    ) {
+        Point valuePoint = SpatialCoordinateTypes.UNSPECIFIED.wkbAsPoint(value);
+        GeoPoint valueGeoPoint = new GeoPoint(valuePoint.getY(), valuePoint.getX());
+
+        Point originPoint = SpatialCoordinateTypes.UNSPECIFIED.wkbAsPoint(origin);
+        GeoPoint originGeoPoint = new GeoPoint(originPoint.getY(), originPoint.getX());
+
+        String originStr = originGeoPoint.getX() + "," + originGeoPoint.getY();
+        String scaleStr = scale.utf8ToString();
+        String offsetStr = offset.utf8ToString();
+
+        return switch (functionType.utf8ToString()) {
+            case "exp" -> new ScoreScriptUtils.DecayGeoExp(originStr, scaleStr, offsetStr, decay).decayGeoExp(valueGeoPoint);
+            case "gauss" -> new ScoreScriptUtils.DecayGeoGauss(originStr, scaleStr, offsetStr, decay).decayGeoGauss(valueGeoPoint);
+            default -> new ScoreScriptUtils.DecayGeoLinear(originStr, scaleStr, offsetStr, decay).decayGeoLinear(valueGeoPoint);
+        };
+    }
+
+    @Evaluator(extraName = "CartesianPoint")
+    static double processCartesianPoint(
+        BytesRef value,
+        @Fixed BytesRef origin,
+        @Fixed double scale,
+        @Fixed double offset,
+        @Fixed double decay,
+        @Fixed BytesRef functionType
+    ) {
+        Point valuePoint = SpatialCoordinateTypes.UNSPECIFIED.wkbAsPoint(value);
+        Point originPoint = SpatialCoordinateTypes.UNSPECIFIED.wkbAsPoint(origin);
+
+        // Euclidean distance
+        double dx = valuePoint.getX() - originPoint.getX();
+        double dy = valuePoint.getY() - originPoint.getY();
+        double distance = Math.sqrt(dx * dx + dy * dy);
+
+        distance = Math.max(0.0, distance - offset);
+
+        return switch (functionType.utf8ToString()) {
+            case "exp" -> {
+                double scaling = Math.log(decay) / scale;
+                yield Math.exp(scaling * distance);
+            }
+            case "gauss" -> {
+                double sigmaSquared = -Math.pow(scale, 2.0) / (2.0 * Math.log(decay));
+                yield Math.exp(-Math.pow(distance, 2.0) / (2.0 * sigmaSquared));
+            }
+            // linear
+            default -> {
+                double scaling = scale / (1.0 - decay);
+                yield Math.max(0.0, (scaling - distance) / scaling);
+            }
+        };
+    }
+
+    @Evaluator(extraName = "Datetime", warnExceptions = { InvalidArgumentException.class, IllegalArgumentException.class })
+    static double processDatetime(
+        long value,
+        @Fixed long origin,
+        @Fixed long scale,
+        @Fixed long offset,
+        @Fixed double decay,
+        @Fixed BytesRef functionType
+    ) {
+        return switch (functionType.utf8ToString()) {
```

Comment:
We're making a lot of conditional logic based on functionType. Would it make sense to create an enum that captures the specific logic for each type, so we don't have to repeat switch expressions on the process methods?

Reply:
[Use enum for DecayFunction instead of switch statements](https://github.com/elastic/elasticsearch/pull/132729/commits/c723309d26d3e20daa8ffee07259c6de79d5b0e2) - Is that what you meant? Definitely looks cleaner!